AdguardTeam/AdGuardHome
# CI revamped and more

First of all thank you for offering AdGuardHome, it's really what I was looking for as a lightweight alternative to Pi-hole.

This PR offers a pretty big redesign of the CI to have a minimum of devops maintenance but also something more versatile. This way, we will be limited to a few dev and production commands.

I would have liked to do several PRs but this was not possible because of the changes that were made. Here is a short summary:

* CI revamped and other enhancements
* GoReleaser for multi-arch binaries
* Switch to GitHub Actions (more versatile) with 2 workflows
* Docker multi-manifest image with buildx
* Go sum fix
* Use tools.go paradigm for external dependencies (packr)
* Use go generate for packr

In order to be able to test my modifications I would first need to activate the GitHub Actions on the repository if it is not already done and also add the secrets as I mention in the changelog below. I'm really hype to work with you on this if your roadmap allows it ;)

Here is what it looks like on my fork: https://github.com/crazy-max/AdGuardHome/actions/runs/112195639

## Changelog

* Add `.dockerignore`
* Global ignore for packr files
* ~~Fix eslint run script~~ (fixed through rebase)
* Add goreleaser config
  * Multi arch binaries
  * Snapcraft
* Only one Dockerfile with multi-platform support through Buildx
* Go sum cleaned
* Use tools.go paradigm for external dependencies (packr)
* Use `go generate` for packr
* Add Open Container Specification labels to Dockerfile
* Switch to GitHub Actions (more versatile) with 3 jobs for build workflow
  * Test
    * 3 platforms (linux,windows,macos) to enhance coverage
    * Cache go modules and node_modules
    * golangci-lint ported
    * eslint
    * Go tests ported
    * Coverage uploaded to Codecov
  * App
    * Cache go modules and node_modules
    * npm ci and build-prod
    * Snapcraft Login (if tag)
    * GoReleaser release or snapshot if not a tag
    * Create GitHub release to publish assets as draft for review (if tag)
  * Docker multi-arch manifest
    * On release: create x.x.x and latest tags
    * On snapshot: crate edge tag
    * Install docker buildx
    * Docker build
    * Docker login (if tag)
    * Docker push (if tag)
    * Image manifest check (if tag)
    * Clear Docker credentials on post (if tag)
* Add simple workflow to fix go.sum if not "tied"
* GitHub Actions secrets have to be added to the repo: `CODECOV_TOKEN`, `SNAPCRAFT_LOGIN`, `DOCKER_USERNAME`, `DOCKER_PASSWORD`

And thanks to buildx (Buildkit), we have a single multi-arch image. Will look like this on DockerHub:

![image](https://user-images.githubusercontent.com/1951866/82643030-25bff100-9c0f-11ea-902d-19b0f8cb46b5.png)


## Local dev commands

* `npm --prefix client ci`
* `npm --prefix client run build-prod`
* `CHANNEL=none goreleaser release --rm-dist --snapshot`
* `docker build -t adguardhome -f ./Dockerfile .`

## TODO

* [X] Add Slack notification
* [ ] Make a review of snapcraft build
